taichi/program/ndarray.cpp

```diff
@@ -15,21 +15,31 @@ Ndarray::Ndarray(Program *prog,
                                 std::end(shape),
                                 1,
                                 std::multiplies<>())),
-      element_size_(data_type_size(dtype)) {
-#ifdef TI_WITH_LLVM
-  LlvmProgramImpl *prog_impl = prog->get_llvm_program_impl();
-  ndarray_alloc_ = prog_impl->allocate_memory_ndarray(nelement_ * element_size_,
-                                                      prog->result_buffer);
+      element_size_(data_type_size(dtype)),
+      device_(prog->get_device_shared()) {
+  ndarray_alloc_ = prog->allocate_memory_ndarray(nelement_ * element_size_,
+                                                 prog->result_buffer);
 
-  data_ptr_ = prog_impl->get_ndarray_alloc_info_ptr(ndarray_alloc_);
+#ifdef TI_WITH_LLVM
+  if (arch_is_cpu(prog->config.arch) || prog->config.arch == Arch::cuda) {
+    // For the LLVM backends, device allocation is a physical pointer.
+    data_ptr_ = prog->get_llvm_program_impl()->get_ndarray_alloc_info_ptr(
+        ndarray_alloc_);
+  }
+#else
+  TI_ERROR("Llvm disabled");
+#endif
 
-  // taichi's own ndarray's ptr points to its |DeviceAllocation| on the
+  // Taichi's own ndarray's ptr points to its |DeviceAllocation| on the
   // specified device. Note that torch-based ndarray's ptr is a raw ptr but
   // we'll get rid of it soon.
   device_allocation_ptr_ = (uint64_t *)&ndarray_alloc_;
-#else
-  TI_ERROR("Llvm disabled");
-#endif
+}
+
+Ndarray::~Ndarray() {
+  if (device_) {
+    device_->dealloc_memory(ndarray_alloc_);
```

Comment:
just for my understanding, how this works with preallocated memory? where we don't deallocate per instance.

Reply:
device_ is nullptr for llvm based backends so this is a no-op for them :D